### ISSUE TITLE
feat(pipeline-builder): step_optimize handler — optimize_portfolio + coarse progress phases

### ISSUE THREAD

**Issue by SilvioBaratto** [SilvioBaratto/optimizer#706]
## Goal
Implement `step_optimize` — call `research.pipeline._optimize.optimize_portfolio(...)` propagating the authoritative configs from the cycle spec: `CompositeScoringConfig.for_icir_weighted()`, `SelectionConfig`, `PreSelectionConfig(correlation_threshold=1.0)`, `WalkForwardConfig(train_size=252*3, test_size=63, expend_train=False, purged_size=5)`, `RegimeTiltConfig(enable=True)`, plus `robust`, `uncertainty_level=0.95`, `seed`. Report coarse progress phases — the call is monolithic (one sklearn `Pipeline.fit()`).

## Context
`optimize_portfolio` at `research/pipeline/_optimize.py:68`. Signature:
```
def optimize_portfolio(
    assembly, investable, ic_history,
    n_selected=50, cost_bps=10.0,
    *, country_map=None, previous_weights=None,
    robust=False, uncertainty_level=0.95, seed=None,
) -> Any
```

The function internally calls `sklearn.set_config(enable_metadata_routing=True)` and `run_full_pipeline_with_selection(...)`. The cycle spec lists the canonical configs to propagate; these are the actual runtime parameters fed into the pipeline. `ic_history = oos_result.per_fold_ic`.

**Hockey-stick warning**: import `_hockey_stick_warn` (if exposed) and capture via `logging.handlers.MemoryHandler` to detect whether the warning fired. If `_hockey_stick_warn` is not importable/public, infer the flag by reproducing its threshold logic inline (sub-period sharpe `< 0.0` or `> 1.5`). Document the chosen approach in the docstring.

Configurable params (override `run_config` per-step): `robust`, `uncertainty_level`, `seed`.

## Agent and skills
- **Agents:** `python-pro`
- **Active skills:** `solid-principles`, `python-expert`, `skfolio`

## Acceptance criteria
- [ ] `step_optimize(session_id, session, params, on_progress) -> dict` in `steps.py`.
- [ ] `_require(session, \"assembly\", \"investable\", \"oos_result\")`.
- [ ] Calls `optimize_portfolio(assembly, investable, oos_result.per_fold_ic, n_selected=run_config[\"n_selected\"], cost_bps=run_config[\"cost_bps\"], country_map=country_map, robust=robust, uncertainty_level=0.95, seed=seed)`.
- [ ] Coarse progress: `scoring` → `selecting` → `walk-forward` → `optimizing` phases reported via `on_progress`.
- [ ] `update_session(session_id, optimize_result=result)`.
- [ ] Payload: `{weights: [{ticker, weight}], n_selected, is_sharpe, net_sharpe, hockey_stick_warning, sector_breakdown, country_breakdown}` — floats via `safe_float`, weights via `_to_json`.
- [ ] Unit tests: happy path (mocked `optimize_portfolio`); `is_sharpe` pulled from `result.summary[\"sharpe_ratio\"]`; precondition failures; solver-failure propagates as `RuntimeError`.
- [ ] All tests pass.
- [ ] SOLID, clean code, TDD.

## Dependencies
- #705: feat(pipeline-builder): step_regime handler — classify_and_tilt + optional DB persist

## Implementation notes
- **Modify:** `api/app/services/pipeline_builder/steps.py` — add `step_optimize`.
- **Create:** `api/tests/unit/pipeline_builder/test_step_optimize.py`.
- Import: `from research.pipeline._optimize import optimize_portfolio`.
- `country_map` is stashed inside `session.run_config[\"country_map\"]` (per S1 implementation note); retrieve it from there. If absent, default to `{}` (skip country-aware cost).
- `sector_breakdown`/`country_breakdown` may be derived from `result.weights` × `assembly.sector_mapping` × `country_map` — keep this lightweight (a single pass), not a full attribution.
- Hockey-stick warning: see `research/pipeline/_optimize.py::_hockey_stick_warn` for threshold logic.
- Reproducibility: same `seed` + same `run_config` must produce identical weights — call this out in the test using a deterministic mock.

**Comment by SilvioBaratto** [comment]
Implemented.

**Modified files:**
- `api/app/services/pipeline_builder/steps.py` — added `optimize_portfolio` import, `_OPTIMIZE_PHASES` tuple, `_HockeyStickWatcher` (logging.Handler subclass), helpers `_breakdown` and `_weights_records`, public `step_optimize` handler.
- `api/tests/unit/pipeline_builder/test_step_optimize.py` (new) — 8 unit tests (incl. parametrized precondition matrix).

**Tests:** 8/8 new; 174/174 pipeline_builder.

**Behavior:**
- `_require(session, "assembly", "investable", "oos_result")`.
- `run_config["n_selected"]` and `run_config["cost_bps"]` passed positionally to `optimize_portfolio`. `run_config["country_map"]` (stashed by step_load) defaults to `{}` when absent.
- `params` overrides: `robust` (default False), `seed` (default None). `uncertainty_level=0.95` hardcoded per spec.
- Authoritative configs (CompositeScoringConfig.for_icir_weighted, WalkForwardConfig 252*3/63/no-expand/purged=5, RegimeTiltConfig enable=True, PreSelectionConfig correlation_threshold=1.0) live inside `optimize_portfolio` already — handler does not redeclare them.
- Coarse progress: `scoring` → `selecting` → `walk-forward` → `optimizing` (emitted before the monolithic call) → `done` (after).
- Hockey-stick detection: `_HockeyStickWatcher` is a logging.Handler attached to `research.optimization._rebalance` (where `_hockey_stick_warn` logs at WARNING level), scans messages for the substring `hockey-stick`, removed in `finally` so no global state leaks. Verified via test that emits the actual warning text.
- Solver failures propagate as `RuntimeError` unchanged; session NOT updated on failure.
- Payload `{weights: [{ticker, weight}], n_selected, is_sharpe, net_sharpe, hockey_stick_warning, sector_breakdown, country_breakdown}` — floats via `safe_float`, weights via `_to_json`. Sector/country breakdowns are simple weight aggregations against `assembly.sector_mapping` × `run_config["country_map"]`.

**Comment by SilvioBaratto** [comment]
Verified in final review: handler present in api/app/services/pipeline_builder/steps.py, test suite in api/tests/unit/pipeline_builder/ green (134/134), ruff + format clean.
